core/src/main/java/org/apache/iceberg/DeleteFileIndex.java

```diff
@@ -96,21 +100,101 @@ private StructLikeWrapper newWrapper(int specId) {
+          limitBySequenceNumber(sequenceNumber, partitionDeletes.first(), partitionDeletes.second()));
+    }
+
+    return matchingDeletes
+        .filter(deleteFile -> canContainDeletesForFile(file, deleteFile, schema))
+        .toArray(DeleteFile[]::new);
+  }
+
+  private static boolean canContainDeletesForFile(DataFile dataFile, DeleteFile deleteFile, Schema schema) {
+    switch (deleteFile.content()) {
+      case POSITION_DELETES:
+        // check that the delete file can contain the data file's file_path
+        Map<Integer, ByteBuffer> lowers = deleteFile.lowerBounds();
+        Map<Integer, ByteBuffer> uppers = deleteFile.upperBounds();
+        if (lowers == null || uppers == null) {
+          return true;
+        }
+
+        Type pathType = MetadataColumns.DELETE_FILE_PATH.type();
+        int pathId = MetadataColumns.DELETE_FILE_PATH.fieldId();
+        ByteBuffer lower = lowers.get(pathId);
+        if (lower != null &&
+            Comparators.charSequences().compare(dataFile.path(), Conversions.fromByteBuffer(pathType, lower)) < 0) {
+          return false;
+        }
+
+        ByteBuffer upper = uppers.get(pathId);
+        if (upper != null &&
+            Comparators.charSequences().compare(dataFile.path(), Conversions.fromByteBuffer(pathType, upper)) > 0) {
+          return false;
+        }
+
+        break;
+
+      case EQUALITY_DELETES:
+        if (dataFile.lowerBounds() == null || dataFile.upperBounds() == null ||
+            deleteFile.lowerBounds() == null || deleteFile.upperBounds() == null) {
+          return true;
+        }
+
+        Map<Integer, ByteBuffer> dataLowers = dataFile.lowerBounds();
+        Map<Integer, ByteBuffer> dataUppers = dataFile.upperBounds();
+        Map<Integer, ByteBuffer> deleteLowers = deleteFile.lowerBounds();
+        Map<Integer, ByteBuffer> deleteUppers = deleteFile.upperBounds();
+
+        for (int id : deleteFile.equalityFieldIds()) {
+          Type type = schema.findType(id);
+          if (!type.isPrimitiveType()) {
+            return true;
+          }
+
+          if (!rangesOverlap(type.asPrimitiveType(),
+              dataLowers.get(id), dataUppers.get(id), deleteLowers.get(id), deleteUppers.get(id))) {
+            return false;
+          }
+        }
+        break;
+    }
+
+    return true;
+  }
+
+  private static <T> boolean rangesOverlap(Type.PrimitiveType type,
+                                           ByteBuffer dataLower, ByteBuffer dataUpper,
+                                           ByteBuffer deleteLower, ByteBuffer deleteUpper) {
+    Comparator<T> comparator = Comparators.forType(type);
+    T low = Conversions.fromByteBuffer(type, dataLower);
+    T high = Conversions.fromByteBuffer(type, dataUpper);
+
+    if (contains(comparator, low, high, Conversions.fromByteBuffer(type, deleteLower))) {
+      return true;
+    }
+
+    if (contains(comparator, low, high, Conversions.fromByteBuffer(type, deleteUpper))) {
+      return true;
```

Comment:
How about the case:  data file1 have a range ['3', '5'],  delete file2 have a range ['2', '8'] ?   I think it should also be overlap but here we will return false ? 

IMO, we should return `comparator.compare(deleteLower, dataUpper) <= 0 && comparator.compare(deleteUpper, dataLower) >=0`  .    That should handle all overlap cases.

Reply:
You're right, I'll take a look at this.

Reply:
Fixed.